core/controllers/editor_test.py

```diff
@@ -605,6 +587,332 @@ def test_state_yaml_handler(self):
 
         self.logout()
 
+    def test_state_yaml_handler_with_no_state_dict_raises_error(self):
+        self.login(self.OWNER_EMAIL)
+        exp_id = 'eid'
+        owner_id = self.get_user_id_from_email(self.OWNER_EMAIL)
+        self.save_new_valid_exploration(exp_id, owner_id)
+
+        response = self.get_html_response(
+            '%s/%s' % (feconf.EDITOR_URL_PREFIX, exp_id))
+        csrf_token = self.get_csrf_token_from_response(response)
+
+        self.post_json(
+            '/createhandler/state_yaml/%s' % exp_id, {},
+            csrf_token=csrf_token, expected_status_int=404)
+
+        self.logout()
+
+    def test_exploration_download_handler_with_invalid_exploration_id(self):
+        self.login(self.OWNER_EMAIL)
+
+        self.get_json(
+            '/createhandler/download/invalid_exploration_id',
+            expected_status_int=404)
+
+        self.logout()
+
+    def test_exploration_download_handler_with_invalid_output_format(self):
+        self.login(self.OWNER_EMAIL)
+        owner_id = self.get_user_id_from_email(self.OWNER_EMAIL)
+        exp_id = 'exp_id1'
+
+        self.save_new_valid_exploration(exp_id, owner_id)
+
+        response = self.get_json(
+            '/createhandler/download/%s?output_format=invalid_output_format'
+            % (exp_id), expected_status_int=400)
+
+        self.assertEqual(
+            response['error'],
+            'Unrecognized output format invalid_output_format')
+
+        self.logout()
+
+
+class ExplorationSnapshotsHandlerTests(test_utils.GenericTestBase):
+
+    def setUp(self):
+        super(ExplorationSnapshotsHandlerTests, self).setUp()
+        self.signup(self.OWNER_EMAIL, self.OWNER_USERNAME)
+
+    def test_get_with_invalid_exploration_id_raises_error(self):
+        self.login(self.OWNER_EMAIL)
+
+        self.get_json(
+            '/createhandler/snapshots/invalid_exploration_id',
+            expected_status_int=404)
+
+        self.logout()
+
+    def test_get_exploration_snapshot_history(self):
+        self.login(self.OWNER_EMAIL)
+        exp_id = 'eid'
+        owner_id = self.get_user_id_from_email(self.OWNER_EMAIL)
+
+        self.save_new_valid_exploration(exp_id, owner_id)
+
+        snapshots = exp_services.get_exploration_snapshots_metadata(exp_id)
+
+        # Patch `snapshots` to use the editor's display name.
+        for snapshot in snapshots:
+            snapshot.update({
+                'committer_id': 'owner'
+                })
```

Comment:
Deindent by 4 to match indentation level of opening braces.

Reply:
Done